specification/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/openshiftclusters/preview/2024-08-12-preview/redhatopenshift.json

```diff
@@ -0,0 +1,2722 @@
+      "type": "object",
+      "properties": {
+        "clientId": {
+          "type": "string"
+        },
+        "principalId": {
+          "type": "string"
+        }
+      }
+    },
+    "ConsoleProfile": {
+      "description": "ConsoleProfile represents a console profile.",
+      "type": "object",
+      "properties": {
+        "url": {
+          "description": "The URL to access the cluster console.",
+          "type": "string",
+          "readOnly": true
+        }
+      }
+    },
+    "Display": {
+      "description": "Display represents the display details of an operation.",
+      "type": "object",
+      "properties": {
+        "provider": {
+          "description": "Friendly name of the resource provider.",
+          "type": "string"
+        },
+        "resource": {
+          "description": "Resource type on which the operation is performed.",
+          "type": "string"
+        },
+        "operation": {
+          "description": "Operation type: read, write, delete, listKeys/action, etc.",
+          "type": "string"
+        },
+        "description": {
+          "description": "Friendly name of the operation.",
+          "type": "string"
+        }
+      }
+    },
+    "EffectiveOutboundIP": {
+      "description": "EffectiveOutboundIP represents an effective outbound IP resource of the cluster public load balancer.",
+      "type": "object",
+      "properties": {
+        "id": {
+          "description": "The fully qualified Azure resource id of an IP address resource.",
+          "type": "string"
+        }
+      }
+    },
+    "EncryptionAtHost": {
+      "description": "EncryptionAtHost represents encryption at host state",
+      "enum": [
+        "Disabled",
+        "Enabled"
+      ],
+      "type": "string",
+      "x-ms-enum": {
+        "name": "EncryptionAtHost",
+        "modelAsString": true
+      }
+    },
+    "FipsValidatedModules": {
+      "description": "FipsValidatedModules determines if FIPS is used.",
+      "enum": [
+        "Disabled",
+        "Enabled"
+      ],
+      "type": "string",
+      "x-ms-enum": {
+        "name": "FipsValidatedModules",
+        "modelAsString": true
+      }
+    },
+    "Identity": {
+      "description": "Identity stores information about the cluster MSI(s) in a workload identity cluster.",
+      "type": "object",
+      "properties": {
+        "type": {
+          "$ref": "#/definitions/ResourceIdentityType"
+        },
+        "principalId": {
+          "type": "string",
+          "readOnly": true
+        },
+        "tenantId": {
+          "type": "string",
+          "readOnly": true
+        },
+        "userAssignedIdentities": {
+          "type": "object",
+          "additionalProperties": {
+            "$ref": "#/definitions/ClusterUserAssignedIdentity"
+          }
```

Comment:
Use of `additionalProperties` is not allowed for properties owned by the service. The only time its ok to use it is when the properties are pass thru (user defined) and not subject to any validations. Please make this is an array or provide an explanation for why you need this.

Reply:
Hi @ms-bogdan-rotaru, thanks for the review. in this case and according to my understanding, the array model is deprecated by the managed identity team: https://eng.ms/docs/products/arm/rbac/managed_identities/msionboardinguserassigned#api-requirements--design that is why we are using a dictionary/map instead which is probably what is resulting in our swagger generation adding the ClusterUserAssignedIdentity as `additionalProperties`. Our API was designed with the MSI/managed identity team's requirements in mind.

If you take a look at our API definition, you can see how we've implemented this requirement: https://github.com/Azure/ARO-RP/pull/3727/files?diff=unified&w=0#diff-c7bfe4dfde8a1a575263edc226847cbab8ac7f60f04193f531a17d73571a3398R438-R458 and how AKS RP has done the same for a similar use case: https://msazure.visualstudio.com/CloudNativeCompute/_git/aks-rp?path=/resourceprovider/server/microsoft.com/containerservice/datamodel/v20240602preview/types.go&version=GBmaster&line=217&lineEnd=247&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

That being said, I'm not totally clear on what you mean by pass thru properties or validations. Can you explain these points more?